# HG changeset patch
# User Rik <rik@octave.org>
# Date 1473475222 25200
#      Fri Sep 09 19:40:22 2016 -0700
# Node ID 6af82fc041a41c914d2aaf57c19c6fdb5ba0275a
# Parent  30002d6568c4aabc5bf795179d567bb751f885ee
Implement Matlab-compatible MEX functions mxAssert and mxAssertS.

* mexproto.h: Define macros mxAssert and mxAssertS which call
mxErrMsgIdAndTxt.  Macros are protected by #if defined (MEX_DEBUG)
so that there is a performance penalty only for debug versions of code (-g).

* mkoctfile.in.cc: When compiling a mex file with -g debug option,
define MEX_DEBUG so that mxAssert and mxAssertS are available.

diff --git a/libinterp/corefcn/mexproto.h b/libinterp/corefcn/mexproto.h
--- a/libinterp/corefcn/mexproto.h
+++ b/libinterp/corefcn/mexproto.h
@@ -277,47 +277,60 @@ extern OCTINTERP_API const char *mxGetFi
                                                          int key_num);
 extern OCTINTERP_API int mxGetFieldNumber (const mxArray *ptr, const char *key);
 
 extern OCTINTERP_API int mxGetString (const mxArray *ptr, char *buf,
                                       mwSize buflen);
 extern OCTINTERP_API char *mxArrayToString (const mxArray *ptr);
 
 /* Miscellaneous.  */
-#if defined (NDEBUG)
+extern OCTINTERP_API mwIndex
+mxCalcSingleSubscript (const mxArray *ptr, mwSize nsubs, mwIndex *subs);
+
+extern OCTINTERP_API size_t mxGetElementSize (const mxArray *ptr);
+
+#if defined (MEX_DEBUG)
+
 #  define mxAssert(expr, msg)                                           \
   do                                                                    \
     {                                                                   \
       if (! expr)                                                       \
         {                                                               \
-          mexPrintf ("Assertion failed: %s, at line %d of file \"%s\".\n%s\n", \
-                     #expr, __LINE__, __FILE__, msg);                   \
+          if (msg != NULL && msg[0] != '\0')                            \
+            mexErrMsgIdAndTxt ("Octave:MEX",                            \
+                               "Assertion failed: %s, at line %d of file \"%s\".\n%s\n", \
+                               #expr, __LINE__, __FILE__, msg);         \
+          else                                                          \
+            mexErrMsgIdAndTxt ("Octave:MEX",                            \
+                               "Assertion failed: %s, at line %d of file \"%s\".\n", \
+                               #expr, __LINE__, __FILE__);              \
         }                                                               \
     }                                                                   \
   while (0)
 
 #  define mxAssertS(expr, msg)                                          \
   do                                                                    \
     {                                                                   \
       if (! expr)                                                       \
         {                                                               \
-          mexPrintf ("Assertion failed at line %d of file \"%s\".\n%s\n", \
-                     __LINE__, __FILE__, msg);                          \
-          abort ();                                                     \
+          if (msg != NULL && msg[0] != '\0')                            \
+            mexErrMsgIdAndTxt ("Octave:MEX",                            \
+                               "Assertion failed at line %d of file \"%s\".\n%s\n", \
+                               __LINE__, __FILE__, msg);                \
+          else                                                          \
+            mexErrMsgIdAndTxt ("Octave:MEX",                            \
+                               "Assertion failed at line %d of file \"%s\".\n", \
+                               __LINE__, __FILE__);                     \
         }                                                               \
     }                                                                   \
   while (0)
+
 #else
 #  define mxAssert(expr, msg)
 #  define mxAssertS(expr, msg)
 #endif
 
-extern OCTINTERP_API mwIndex
-mxCalcSingleSubscript (const mxArray *ptr, mwSize nsubs, mwIndex *subs);
-
-extern OCTINTERP_API size_t mxGetElementSize (const mxArray *ptr);
-
 #if defined (__cplusplus)
 }
 #endif
 
 #endif
 
diff --git a/src/mkoctfile.in.cc b/src/mkoctfile.in.cc
--- a/src/mkoctfile.in.cc
+++ b/src/mkoctfile.in.cc
@@ -578,16 +578,22 @@ main (int argc, char **argv)
           std::cerr << "mkoctfile: unrecognized argument " << arg << std::endl;
           return 1;
         }
 
       if (! file.empty () && octfile.empty ())
         octfile = file;
     }
 
+  if (output_ext ==  ".mex"
+      && vars["ALL_CFLAGS"].find ("-g") != std::string::npos) 
+    {
+      defs += " -DMEX_DEBUG";
+    }
+
   if (link_stand_alone)
     {
       if (! outputfile.empty ())
         output_option = "-o " + outputfile;
     }
   else
     {
       if (! outputfile.empty ())
